### ISSUE TITLE
Allow to dynamically depend on plugins based on current app flavor

### ISSUE THREAD

**Issue by TheOneWithTheBraid** [flutter/flutter#80940]
## Use case

Many Flutter applications rely on native dependencies. Often, these native dependencies use the Google Play Services, Huawei App Services or Amazon Fire APIs on Android, make use of Vendor specific APIs or rely on special APIs (such as root features). Unfortunately, it is not possible to publish an application relying on certain APIs in all different marketplaces.

We are hence looking for a way to dynamically in- and exclude certain dependencies for certain build flavors.

E.g. `myapp` depends on the fictional package `accesses_restricted_api` whereas `accesses_restricted_api` makes use of a native API being only available on app marketplace 1 and being disallowed on app marketplace 2 and app marketplace 3 as well as unwanted in regard of the marketplace's philosophy in app marketplace 4.

We are hence looking for a way to blacklist/remove the package (or, to be clear a certain class name) in one of our build flavors.

To be clear for those people usually only publishing in the major app marketplaces: it's currently not possible to use Flutter's own capabilities to publish an app making use of one of the various native Google Service Framework-dependent plugins in Google Play Store, Huawei AppGallery, Amazon Appstore and F-Droid.

## Workarounds tried

We tried to fork the fictional `acces_restricted_api` and create package-specific android build flavors too. Unfortunately, Flutter requires the native class to be present in `android/src/**main**` even if we would like to have different implementations for the different build flavors (e.g. in `android/src/**libre**`, `android/src/**marketplace1**` and `android/src/**marketplace2**`).

As this does not work, we now work around this by maintaining different branches for each app marketplace we target. This is a tremendous amount of work as every change needs to be integrated into four different code bases.

## Proposal

We can imagine two different implementations:

- ether allow flavor-specific implementations for native code on Android
- or provide a way to remove / blacklist a dependency at compile time in case a certain flavor is used

**Comment by blasten** [comment]
Do you own these plugins?  If yes, a Flutter Android plugin is a Gradle subproject, so you could do something like this: https://stackoverflow.com/a/18213259, by referring to the plugin's subproject: `project('acces_restricted_api')`.

**Comment by neokree** [comment]
Hello, In my case most of the time these plugins are third party, like GoogleMap and MapBox sdks. 
I also think that this is not only an android problem, but also iOS, web etc. 

A way to have "flavors" directly integrated in Flutter's `pubspec.yaml`, making each flavour to have a list of specific dependencies, added to the more general ones, would be the perfect solution for me. Surely we would also need different entry points (main.dart) that only uses these different dependencies, more or less like plugins already do for `dart:io` and `dart:js`, but without having to use multiple flutter projects

**Comment by blasten** [comment]
have you thought about creating your own CLI tool to achieve this behavior? We will be more than happy to review a proposed design.

**Comment by neokree** [comment]
Yes I thought it before, but from my understanding a CLI tool is not feasible at this time because at best it would require to use a cli command every time we want to switch from one flavor to another. This means that the pubspec.yaml can't be stored in version control, you would need to have several pubspec.yaml, one for each flavor, which means shared dependencies would be separated from specific flavor ones.

Also having it in `flutter` directly would allow this process to be automated as we already need to specify the `--flavor` option when we build (either on debug or on release). 

We could make a tool if flutter would allow us to use custom procedures in the build process, like gradle on android do

**Comment by TheOneWithTheBraid** [comment]
I think this issue was unintentionally closed. The information requested by the bot does not apply here. I would be happy in case someone could re-open this issue.

**Comment by TheOneWithTheBraid** [comment]
> Do you own these plugins?

Some yes, some no.

At the moment, the workaround is a hacky combination of patch files and branches we apply and rebase in CI in order to use different native plugins.

**Comment by rostopira** [comment]
Seems like a duplicate of #46979
